templates/amm.cpp.in

```diff
@@ -319,50 +320,59 @@ double calculate_amm_impl(const @ModelName@_mass_eigenstates& model, const softs
       return std::numeric_limits<double>::quiet_NaN();
    }
 
-   double val =
-      // vector form factor
-      1./2.*(form_factors[2] + form_factors[3]).real()
-      // reinstate the mass that was factored out in definition of form factor
-      * context.mass<Lepton>({@extraIdxUsageNoComma@})
-      // factor out e/(2*lepton pole mass)
-      / (unit_charge(context_base{model})/(2.*lepton_pole_mass<Lepton>(qedqcd@extraIdxUsage@)))
-      // definition of photon momentum flow for g-2 is opposite than for form factors
-      * (-1.);
+   double val = 0.;
+
+   // 1L
+   if (settings.get(Spectrum_generator_settings::calculate_amm) > 0.5) {
+      val +=
+         // vector form factor
+         1./2.*(form_factors[2] + form_factors[3]).real()
+         // reinstate the mass that was factored out in definition of form factor
+         * context.mass<Lepton>({@extraIdxUsageNoComma@})
+         // factor out e/(2*lepton pole mass)
+         / (unit_charge(context_base{model})/(2.*lepton_pole_mass<Lepton>(qedqcd@extraIdxUsage@)))
+         // definition of photon momentum flow for g-2 is opposite than for form factors
+         * (-1.);
+   }
 
-   // add 2-loop QED logarithms
-   val *= 1. + get_QED_2L<Lepton>(context, qedqcd, lepton_pole_mass<Lepton>(qedqcd@extraIdxUsage@));
+   // 2L QED
+   if (settings.get(Spectrum_generator_settings::calculate_amm) > 1.25) {
+      // add 2-loop QED logarithms
+      val *= 1. + get_QED_2L<Lepton>(context, qedqcd, lepton_pole_mass<Lepton>(qedqcd@extraIdxUsage@));
+   }
 
-   std::complex<double> valBarZee {0., 0.};
+   // Barr-Zee
+   if (settings.get(Spectrum_generator_settings::calculate_amm) > 1.75) {
+      std::complex<double> valBarZee {0., 0.};
```

Comment:
The Barr-Zee loop functions always return a `double`. So `valBarZee` should be made a `double`, too.

Reply:
Note: In `FlexibleSUSY.m` line 2528 we have:
~~~.cpp
            barZee = barZee <>
               "valBarZee += std::complex<double> " <> ToString @ N[
                  ReIm @ CXXDiagrams`ColourFactorForIndexedDiagramFromGraph[
               CXXDiagrams`IndexDiagramFromGraph[diagrams[[i,j]], graphs[[i]]],
                  graphs[[i]]
                ], 16] <> " * " <>
                ToString @ AMM`CXXEvaluatorForDiagramFromGraph[diagrams[[i,j]], graphs[[i]]] <>
                "::value({" <> leptonIndex <> "}, context, qedqcd);\n"
~~~
Here, the prefactor of the Barr-Zee diagram is chosen to be a complex number. This complex number comes from ``CXXDiagrams`ColourFactorForIndexedDiagramFromGraph``. I believe this color factor is always purely real.

Reply:
Because in general a colour factor for a diagram is complex. What we could do here is to write it as real but check if we don't generate an imaginary part. Something like
```mathematica
If[Im[x] == 0, x, Print["Error: Colour prefactor of a Barr-Zee diagram should be real!"]; Quit[1]]
```

Reply:
Ok, I see. Then I'd propose the following: We keep the code as it is (with complex color factors etc.). However, since $a_\mu$ is defined to be a real part, we simply write:
~~~.cpp
val += valBarZee.real();
~~~
and remove the check that `is_zero(valBarZee.imag())`.

Reply:
Done but slightly differently than you wanted. See bc37c3e37ba7f23a4df96e71871d94ec67ca2797.